substrate/frame/assets/src/lib.rs

```diff
@@ -1683,6 +1725,104 @@ pub mod pallet {
 			Self::deposit_event(Event::<T, I>::Blocked { asset_id: id, who });
 			Ok(())
 		}
+
+		/// Revoke irreversibly Owner, Issuer, Admin, and Freezer privileges of the asset. Freeze
+		/// the metadata.
+		///
+		/// Warning: this action is irreversible.
+		///
+		/// Origin must be either Signed and the sender should be the Owner of the asset `id` or
+		/// `ForceOrigin`.
+		///
+		/// If origin is the owner: The deposit of the asset details and metadata will be
+		/// unreserved then withdrawn and sent to `DepositDestinationOnRevocation`. If no metadata
+		/// was set then the new deposit for the new metadata will be withdrawn from the owner.
+		///
+		/// If origin is `ForceOrigin`: The deposit of owner is unreserved.
+		///
+		/// Asset must be in `Live` status.
+		///
+		/// - `id`: The identifier of the asset.
+		///
+		/// Emits `OwnerAndTeamRevoked` and `MetadataSet`.
+		///
+		/// Weight: `O(1)`
+		#[pallet::call_index(32)]
+		pub fn revoke_all_privileges(
+			origin: OriginFor<T>,
+			id: T::AssetIdParameter,
+		) -> DispatchResult {
+			let maybe_owner_origin = match T::ForceOrigin::try_origin(origin) {
+				Ok(_) => None,
+				Err(origin) => Some(ensure_signed(origin)?),
+			};
+			let id: T::AssetId = id.into();
+			let mut asset = Asset::<T, I>::get(id.clone()).ok_or(Error::<T, I>::Unknown)?;
+
+			if let Some(owner) = maybe_owner_origin.as_ref() {
+				ensure!(asset.owner() == Some(&owner), Error::<T, I>::NoPermission);
+			}
+			ensure!(asset.status == AssetStatus::Live, Error::<T, I>::IncorrectStatus);
+
+			let old_metadata = Metadata::<T, I>::get(&id);
+
+			let deposit_amount = asset.deposit + old_metadata.deposit;
+
+			let fee_imbalance = if let Some(owner) = maybe_owner_origin {
+				let new_metadata_deposit =
+					Self::calc_metadata_deposit(&old_metadata.name, &old_metadata.symbol);
+
+				// NOTE: if new deposit is less than old deposit then the excess is not refunded.
+				let metadata_deposit_diff =
+					new_metadata_deposit.saturating_sub(old_metadata.deposit);
+
+				with_storage_layer::<_, DispatchError, _>(|| {
+					T::Currency::unreserve(&owner, deposit_amount);
+					let imbalance = T::Currency::withdraw(
+						&owner,
+						deposit_amount + metadata_deposit_diff,
+						WithdrawReasons::FEE,
+						ExistenceRequirement::KeepAlive,
+					)?;
+					Ok(imbalance)
+				})?
+			} else if let Some(owner) = asset.owner() {
+				T::Currency::unreserve(owner, deposit_amount);
```

Comment:
I thought it could only happen if some of the reserved asset was slashed.
Is there any other reason?
The idea here is that the deposit is given back to the account. If not enought money is reserved then I don't think we should mint some money into the account.

Reply:
Yeah it shouldn't happen, and even if it does we shouldn't mint anything to the account.

But you are passing the imbalance to `T::DepositDestinationOnRevocation::on_unbalanced(fee_imbalance);`, and if some doesn't actually get unreserved then there will be a diff between what gets unreserved and what gets passed into the handler.

Reply:
This branch is in the case if the call comes from `ForceOrigin`.
In this case we don't ask the owner to fund the deposit of the asset. I considered that it is `ForceOrigin` responsibility to ensure the storage is not abused.

So if `ForceOrigin` forces to revoke all privileges on an asset then the owner just get its deposit back and 0 is given to `T::DepositDestinationOnRevocation`.